server/kv.go

```diff
@@ -655,3 +440,203 @@ func (store store) OneTimeLoadOauth1aTemporaryCredentials(mmUserId string) (*OAu
+}
+
+func (store *store) LoadInstances() (*Instances, error) {
+	kv := kvstore.NewStore(kvstore.NewPluginStore(store.plugin.API))
+	vs, err := kv.ValueIndex(keyInstances, &instancesArray{}).Load()
+	if err != nil {
+		return nil, err
+	}
+	return &Instances{
+		ValueSet: vs,
+	}, nil
+}
+
+func (store *store) StoreInstances(instances *Instances) error {
+	kv := kvstore.NewStore(kvstore.NewPluginStore(store.plugin.API))
+	return kv.ValueIndex(keyInstances, &instancesArray{}).Store(instances.ValueSet)
+}
+
+func (store *store) UpdateInstances(updatef func(instances *Instances) error) error {
+	instances, err := store.LoadInstances()
+	if err == kvstore.ErrNotFound {
+		instances = NewInstances()
+	} else if err != nil {
+		return err
+	}
+	err = updatef(instances)
+	if err != nil {
+		return err
+	}
+	return store.StoreInstances(instances)
+}
+
+// MigrateV2Instances migrates instance record(s) from the V2 data model.
+//
+// - v2keyKnownJiraInstances ("known_jira_instances") was stored as a
+//   map[string]string (InstanceID->Type), needs to be stored as Instances.
+//   https://github.com/mattermost/mattermost-plugin-jira/blob/885efe8eb70c92bcea64d1ced6e67710eda77b6e/server/kv.go#L375
+// - v2keyCurrentJIRAInstance ("current_jira_instance") stored an Instance; will
+//   be used to set the default instance.
+// - The instances themselves should be forward-compatible, including
+// 	 CurrentInstance.
+func (store *store) MigrateV2Instances() error {
+	_, err := store.plugin.instanceStore.LoadInstances()
+	if err != kvstore.ErrNotFound {
+		return err
+	}
+
+	// The V3 "instances" key does not exist. Migrate.
+	data, appErr := store.plugin.API.KVGet(v2keyKnownJiraInstances)
+	if appErr != nil {
+		return appErr
+	}
+	v2instances := map[string]string{}
+	if len(data) != 0 {
+		err = json.Unmarshal(data, &v2instances)
+		if err != nil {
+			return err
+		}
+	}
+	instances := NewInstances()
+	for k, v := range v2instances {
+		instances.Set(&InstanceCommon{
+			PluginVersion: manifest.Version,
+			InstanceID:    types.ID(k),
+			Type:          InstanceType(v),
+		})
+	}
+
+	instance, err := store.loadInstance(v2keyCurrentJIRAInstance)
+	if err != nil && err != kvstore.ErrNotFound {
+		return err
+	}
+	if instance != nil {
+		instance.Common().InstanceID = types.ID(instance.GetURL())
+		instances.Set(instance.Common())
+		err = store.StoreInstance(instance)
+		if err != nil {
+			return err
+		}
+
+		if instances.Len() > 1 {
+			instances.SetDefault(instance.GetID())
+		}
+	}
+
+	err = store.StoreInstances(instances)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// MigrateV2User migrates a user record from the V2 data model if needed. It
+// returns an up-to-date User object either way.
+func (p *Plugin) MigrateV2User(mattermostUserID types.ID) (*User, error) {
+	user, err := p.userStore.LoadUser(mattermostUserID)
+	if err != kvstore.ErrNotFound {
+		// return the existing key (or error)
+		return user, err
```

Comment:
I'm experiencing an issue here. The user key is not in the kvstore, yet no error is bubbled up from https://github.com/levb/mattermost-plugin-jira/blob/545ff584f0797b5015b4669618997590c30e27c2/server/kv.go#L113 which is necessary for continuing the migration

Reply:
Changing this to `if err != kvstore.ErrNotFound && user.MattermostUserID != "" {` seems to work in the meantime for my testing.